Editor: Fix flash of unstyled content when loading editor

stephanethomas opened the pull request:
This pull request addresses https://github.com/Automattic/wp-calypso/issues/18703 by fixing a [flash of unstyled content](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) that would otherwise happen the first time the editor is loaded, something quite noticeable with slow connections:
 
![posteditor-flashofsidebar](https://user-images.githubusercontent.com/8658164/31400329-197a85c2-ade7-11e7-9ee2-09899521a5f6.gif)

The solution chosen consists in modifying the [Webpack loader](https://github.com/Automattic/wp-calypso/blob/addc1ccfd359d77e92e204211efac51e4c5721f7/server/bundler/loader.js) that handles [code splitting](https://github.com/Automattic/wp-calypso/tree/addc1ccfd359d77e92e204211efac51e4c5721f7/server/bundler) in order to pre-load the CSS bundle of a section at the same time its Javascript bundle is pre-loaded.

#### Testing instructions
 
Depending on your Internet connection, you may want to set [throttling](https://developers.google.com/web/tools/chrome-devtools/network-performance/reference#throttling) to `Slow 3G` - or something even slower - in the `Network` panel in the Chrome devtools. Finally do not hesitate to repeat the following steps with any [RTL language](http://calypso.localhost:3000/me/account) such as Hebrew:

1. Run `git checkout fix/editor-fouc` and start your server, or open a [live branch](https://calypso.live/?branch=fix/editor-fouc)

##### Scenario 1

2. Open the [`Reader` page](http://calypso.localhost:3000/) in a new tab
3. Click the `Write` button in the top bar
4. Assert that the editor loads with the correct styles applied right away

##### Scenario 2

2. Open the [`My Site` page](http://calypso.localhost:3000/stats/day) in a new tab
3. Click the `Add` button next to the `Sites Pages` or `Blog Posts` menu item in the sidebar
4. Assert that the editor loads with the correct styles applied right away

##### Scenario 3

2. Open the [`Blog Posts` page](http://calypso.localhost:3000/posts) in a new tab
3. Click the `Start a Post` button that appears when you haven't published any posts yet
4. Assert that the editor loads with the correct styles applied right away

##### Scenario 4

2. Open the [`Page Drafts` page](http://calypso.localhost:3000/pages/drafts) in a new tab
3. Click the `Start a Page` button that appears when you haven't published any posts yet
4. Assert that the editor loads with the correct styles applied right away

##### Scenario 5

2. Open the [`Blog Posts` page](http://calypso.localhost:3000/posts) in a new tab
3. Click the `Edit` action below any exising post
4. Assert that the editor loads with the correct styles applied right away

##### Scenario 6

2. Open the [`Site Pages` page](http://calypso.localhost:3000/pages) in a new tab
3. Click the `Edit` menu item that appears when you click the three dots in front of each page
4. Assert that the editor loads with the correct styles applied right away

##### Scenario 7

2. Open the [`Reader` page](http://calypso.localhost:3000/) in a new tab
3. Click the first post in the list
4. Move your mouse over the sharing icon located at the bottom of the post
5. Assert that the layout of the page doesn't change

#### Additional notes

You may want to review changes with [whitespaces ignored](https://github.com/Automattic/wp-calypso/pull/19396/files?w=1), or [commit by commit](https://github.com/Automattic/wp-calypso/pull/19396/commits).

This pull request changes the code generated by the Webpack loader from [this](https://gist.github.com/stephanethomas/36e15a7fd3f5281fe92a2ef782eff4d5/b2b8149479145b7a9e773d635170dd818a78deac) to [this](https://gist.github.com/stephanethomas/36e15a7fd3f5281fe92a2ef782eff4d5/491688964714a5811a94f70223cf2cccf4ee2a99). You can compare those changes by downloading these two revisions, and using your favorite diff tool.

Props to @jeremeylduvall for suggesting preloading the CSS bundles when hovering the button.

#### Reviews
 
- [ ] Code
- [ ] Product

stephanethomas commented on client/components/empty-content/index.jsx:
```diff
@@ -59,6 +64,8 @@ class EmptyContent extends Component {
 					className="empty-content__action button is-primary"
 					href={ this.props.actionURL }
+					onMouseEnter={ this.props.actionHoverCallback }
+					onTouchStart={ this.props.actionHoverCallback }
```
I've included this [`touchstart` event](https://developer.mozilla.org/en-US/docs/Web/Events/touchstart) to be consistent with what we've been doing in [other parts of our code](https://github.com/Automattic/wp-calypso/blob/6300bcfa7aae99ff2f0a349af74313bc9029bfa7/client/blocks/daily-post-button/index.jsx#L163-L164).

stephanethomas commented on client/layout/masterbar/publish.jsx:
```diff
@@ -94,7 +94,8 @@ class MasterbarItemNew extends React.Component {
 				</MasterbarItem>
-				<AsyncLoad require="layout/masterbar/drafts" />
+
+				<AsyncLoad require="layout/masterbar/drafts" placeholder={ null } />
```
This hides a loading placeholder that would otherwise appear in the top bar. This is not related to the original issue though.

gwwar replied:
If you have time do you mind adding a quick shot of that to the summary?

stephanethomas replied:
Haha, I originally put it there but then figured it would make more sense here. I'll add it back.

stephanethomas replied:
A heads-up that this fix was deployed with https://github.com/Automattic/wp-calypso/pull/19558.

stephanethomas commented:
Here is a quick overview of the code generated by the Webpack loader:

```javascript
var config = require( 'config' ),
  page = require( 'page' ),
  React = require( 'react' ),
  activateNextLayoutFocus = require( 'state/ui/layout-focus/actions' ).activateNextLayoutFocus,
  LoadingError = require( 'layout/error' ),
  controller = require( 'controller' ),
  restoreLastSession = require( 'lib/restore-last-path' ).restoreLastSession,
  preloadHub = require( 'sections-preload' ).hub,
  switchCSS = require( 'lib/i18n-utils/switch-locale' ).switchCSS,
  debug = require( 'debug' )( 'calypso:bundler:loader' );

var _loadedSections = {};

function preload( sectionName ) {
  var loadCSS = function( sectionName, cssUrls ) {
    var url = cssUrls.ltr;

    if ( typeof document !== 'undefined' && document.documentElement.dir === 'rtl' ) {
      url = cssUrls.rtl;
    }

    switchCSS( 'section-css', url );
  };

  switch ( sectionName ) {
    case "sites":
      debug( 'Pre-loading Javascript for "sites" section' );

      return require.ensure( [], function() {}, "sites" );

    [...]
    
    case "woocommerce":
      debug( 'Pre-loading Javascript and CSS for "woocommerce" section' );

      loadCSS( 'section-css', {
        "ltr": "/calypso/sections/woocommerce.css?v=e49582d7f4",
        "rtl": "/calypso/sections-rtl/woocommerce.rtl.css?v=bd1dc35294"
      } );

      return require.ensure( [], function() {}, "woocommerce" );

    case "wp-job-manager":
      debug( 'Pre-loading Javascript for "wp-job-manager" section' );

      return require.ensure( [], function() {}, "wp-job-manager" );

    case "wp-super-cache":
      debug( 'Pre-loading Javascript for "wp-super-cache" section' );

      return require.ensure( [], function() {}, "wp-super-cache" );

    case "zoninator":
      debug( 'Pre-loading Javascript for "zoninator" section' );

      return require.ensure( [], function() {}, "zoninator" );
  }
}

preloadHub.on( 'preload', preload );

module.exports = {
  get: function() {
    return [
      {
        "name": "sites",
        "paths": [
          "/sites"
        ],
        "module": "my-sites",
        "group": "sites",
        "secondary": true
      },
      
      [...]
      
      {
        "name": "zoninator",
        "paths": [
          "/extensions/zoninator"
        ],
        "settings_path": "/extensions/zoninator",
        "module": "zoninator",
        "group": "sites",
        "secondary": true,
        "envId": [
          "development",
          "wpcalypso",
          "stage"
        ]
      }
    ];
  },
  load: function() {
    page( /^\/sites(\/.*)?$/, function( context, next ) {
      var envId = undefined;
      if ( envId && envId.indexOf( config( "env_id" ) ) === -1 ) {
        return next();
      }
      if ( _loadedSections[ "my-sites" ] ) {
        controller.setSection( {"name":"sites","paths":["/sites"],"module":"my-sites","group":"sites","secondary":true} )( context );
        context.store.dispatch( activateNextLayoutFocus() );
        return next();
      }
      if ( config.isEnabled( "restore-last-location" ) && restoreLastSession( context.path ) ) {
        return;
      }
      context.store.dispatch( { type: "SECTION_SET", isLoading: true } );
      require.ensure( [],
        function( require ) {
          context.store.dispatch( { type: "SECTION_SET", isLoading: false } );
          controller.setSection( {"name":"sites","paths":["/sites"],"module":"my-sites","group":"sites","secondary":true} )( context );
          if ( ! _loadedSections[ "my-sites" ] ) {
            require( "my-sites" )( controller.clientRouter );
            _loadedSections[ "my-sites" ] = true;
          }
          context.store.dispatch( activateNextLayoutFocus() );
          next();
        }, function onError( error ) {
          if ( ! LoadingError.isRetry() ) {
            console.warn( error );
            LoadingError.retry( "sites" );
          } else {
            console.error( error );
            context.store.dispatch( { type: "SECTION_SET", isLoading: false } );
            LoadingError.show( "sites" );
          }
        }, "sites"
      );
    } );
    
    [...]
  }
};
```

gwwar commented:
Thanks for tackling this! Here's another common scenario:
- Visit the posts or page list directly. `/posts/{siteSlug}`, `/pages/{siteSlug}`
- Click "Edit" on a published post

gwwar commented on client/sections-preload.js:
```diff
@@ -25,11 +25,11 @@ export const hub = {};
-export function preload( section ) {
-	hub.emit( 'preload', section );
+export function preload( sectionName ) {
+	hub.emit( 'preload', sectionName );
```
Is the rename here to clarify that the sectionName is a string?

stephanethomas replied:
Yes, exactly. That's part of [this change](https://github.com/Automattic/wp-calypso/pull/19396/commits/cee500a1bd097b51d1965046765debec9a9714c3).

gwwar commented on server/bundler/loader.js:
```diff
@@ -80,50 +99,51 @@ function getRequires( sections ) {
-		'} );\n'
+		envIdString = JSON.stringify( section.envId );
+
+	const result = [
```
Could you back out the const result assignment? It's making reading this diff a bit difficult

gwwar replied:
I would recommend changes like that for a separate janitorial PR

yurynix replied:
I'm also in favor of not touching that as part of that PR too much...
We have an ongoing PR of refactoring this already, maybe you can comment/commit [there](https://github.com/Automattic/wp-calypso/pull/18383)?

Tug replied:
I agree, we should differentiate refactoring PRs from feature/bugfix ones. A refactoring PR should take into account existing efforts to improve the code and there is at least one as Yury mentioned!

@stephanethomas could you undo all changes that aren't needed for that patch?

nylen replied:
I disagree with this for a few reasons, though of course "it depends" what the right solution is.

- In practice it's unlikely that we will actually follow up with said refactoring, so it's good to go ahead and refresh code that you are touching (as long as the changes are not too severe).
- Without retouching modified lines to conform to current coding standards, we will get lint errors and future changes from `prettier` auto-formatting.

In the case of this PR, reviewing it with the [`?w=1` argument](https://github.com/Automattic/wp-calypso/pull/19396/files?w=1) makes the actual changes really obvious.

stephanethomas replied:
What James said :).

Amber doesn't have time for refactoring these days so quite frankly it would never happen. I know there is another attempt to refactor this part of the code, but it's a pull request started in September and it's still in progress, with no ETA.

I'm happy to revert [to using `var`](https://github.com/Automattic/wp-calypso/pull/19396/commits/6e902771b54bf997183f6cb2a2e715b3ab2c941c) for declaring variables if the [diff with whitespaces ignored](https://github.com/Automattic/wp-calypso/pull/19396/files?w=1) is not enough. Let me know if that's a blocker. Note I've increased the indentation of the generated code here - something that should be part of this pull request - so the [default diff](https://github.com/Automattic/wp-calypso/pull/19396/files) won't be that different from what you currently see.

Tug replied:
In general I would agree, a refactor is a nice bonus usually, but in this case I don't think this refactor improves the code much, it's a personal syntax preference change in my opinion (`` vs `[''].join( '\n' )` ).

The work started in https://github.com/Automattic/wp-calypso/pull/18383 however is more than a syntactic change, it refactors the whole file, improves the logic and I'm pretty sure we can have it merged soon.

*TLDR:* I'm not a fan of refactors that only slightly improves the code and make the review process harder.

gwwar replied:
> In practice it's unlikely that we will actually follow up with said refactoring 

😢 Splitting out janitorials makes things faster all around. They're easy to setup, and fast to review. If you don't have time for refactoring please keep these commits out of PRs that change behavior. It's an undue burden on the reviewer as it is much more difficult to read the diff and catch issues.

gwwar commented:
cc @megsfulton want to help test your favorite bug? 😆

gwwar commented:
Awesome work @stephanethomas! I'm so excited for this fix.✨  This tests well for your stated scenarios, though I'd love to see the edit case too from the post list. I tested this locally + in a docker container.

Let's get some more 👀  on the loader changes. Perhaps we could summon @mtias, @blowery, @ockham, @mcsf or @samouri

nylen commented:
cc @alisterscott, we were discussing this one over at https://github.com/Automattic/wp-calypso/pull/18785

stephanethomas commented:
> Here's another common scenario:
> 
>     Visit the posts or page list directly. /posts/{siteSlug}, /pages/{siteSlug}
>    Click "Edit" on a published post

Thanks for the review @gwwar! I've updated the testing instructions with this more scenarios. Some of them were not correctly handled, I'm in the process of fixing that.

gwwar commented:
> Some of them were not correctly handled, I'm in the process of fixing that.

Sounds great! Give me another ping when you're ready for a second round of testing.

yurynix commented:
I've pushed a little commit to fix a tiny issue.
Overall this looks good to me and ready to go.

stephanethomas commented:
Thanks for reviewing it Yury. However I'm afraid it's still in progress, since there are some scenarios that are not correctly handled yet.

gwwar commented:
> However I'm afraid it's still in progress, since there are some scenarios that are not correctly handled yet.

Are the other scenarios blocking? It would be good to get a partial fix in otherwise.

alisterscott commented:
I think this is a big improvement on current behaviour so can we merge this one and work on incremental improvements in subsequent PRs?

I ran the editor (post and page) e2e tests against this branch and all passed ✅

stephanethomas commented:
Sorry for the delay, but we've been busy with tasks with higher priority lately.

>I ran the editor (post and page) e2e tests against this branch and all passed

I've rebased and restarted Canary tests, but three unit tests are failing now. I'm going to investigate.

> Are the other scenarios blocking? It would be good to get a partial fix in otherwise.

I've added a seventh scenario, which is indeed a regression quite noticeable.

alisterscott commented:
> I've rebased and restarted Canary tests, but three unit tests are failing now. I'm going to investigate.

They are failing because they rely upon https://calypso.live/?branch=fix/editor-fouc which isn't working (and I don't know why). So I ran them against this branch locally and all were fine.

alisterscott commented:
I tried the latest revision locally and I can't get npm start to work (after a `clean` and `distclean`):

```
/Users/alisterscott/Projects/a8c/wp-calypso/build/webpack:/client/lib/route/add-query-args.js:16
		throw new Error( 'addQueryArgs expects the second argument to be a string.' );
        ^
Error: addQueryArgs expects the second argument to be a string.
```

gwwar commented:
> I tried the latest revision locally and I can't get npm start to work (after a clean and distclean):

This branch needs to be rebased with master. It's still on Node v6 when we moved to v8

stephanethomas commented:
> This branch needs to be rebased with master. It's still on Node v6 when we moved to v8

I did rebase thus pull request after the Node upgrade. I'm not sure what happened, I'll give it another try.

stephanethomas commented on client/lib/i18n-utils/switch-locale.js:
```diff
@@ -54,16 +54,29 @@ export default function switchLocale( localeSlug ) {
+		callback();
+
+		return;
+	}
```
This is to handle the case where this function is called in a very short time interval, which sometimes leads to injecting the same `<link />` element twice. This particular operation indeed isn't instantaneous, and as such the test below that checks for the presence of this element will fail while one is already in the process of being injected in the DOM.

Tug replied:
Oh I see, because the `id` attribute is set asynchronously... I think we should change this to set it synchronously by default except when it already exists.

stephanethomas commented:
Everything has been fixed, I've tested it again. This is good for a last review.

apeatling commented:
@gwwar Would you be able to give this one a final thumbs up? Thanks.

gwwar commented:
LGTM :shipit:

stephanethomas commented:
Thanks for looking at it @gwwar!
